yarn-project/foundation/src/json-rpc/client/safe_json_rpc_client.ts

```diff
@@ -31,24 +62,100 @@ export function createSafeJsonRpcClient<T extends object>(
 ): T {
   const fetch = config.fetch ?? defaultFetch;
   const log = config.log ?? createLogger('json-rpc:client');
-  const { useApiEndpoints = false, namespaceMethods = false } = config;
+  const { namespaceMethods = false, batchWindowMS = DEFAULT_BATCH_WINDOW_MS } = config;
 
   let id = 0;
+  let sendBatchTimeoutHandle: NodeJS.Timeout | undefined;
+  let queue: Array<{ request: JsonRpcRequest; deferred: PromiseWithResolvers<JsonRpcResponse> }> = [];
+
+  const sendBatch = async () => {
+    if (sendBatchTimeoutHandle !== undefined) {
+      clearTimeout(sendBatchTimeoutHandle);
+      sendBatchTimeoutHandle = undefined;
+    }
+
+    const rpcCalls = queue;
+    queue = [];
+
+    if (rpcCalls.length === 0) {
+      return;
+    }
+
+    log.debug(`Executing JSON-RPC batch of size: ${rpcCalls.length}`, {
+      methods: rpcCalls.map(({ request }) => request.method),
+    });
+    const { response, headers } = await fetch(
+      host,
+      rpcCalls.map(({ request }) => request),
+    );
```

Comment:
Don't we need to try/catch this, and if it fails, reject all deferreds? Otherwise they'll just remain hanging.

Reply:
Also, do we need to handle the case where the batch `response` has a top-level `error`?

Reply:
> Don't we need to try/catch this, and if it fails, reject all deferreds? Otherwise they'll just remain hanging.

🙈 

> Also, do we need to handle the case where the batch response has a top-level error?

This is handled by this if statement
```
    if (!Array.isArray(response) || response.length !== rpcCalls.length) {
```